### PR TITLE
Use temporary file to avoid out-of-memory when receiving big chunks.

### LINKED ISSUE #218: Pushing a large repo with http leads to crash
I have no reproducible environment right now, but I used to host an AOSP (Android Open Source Project) repository and `git push` with http can crash it reliably.

(A crashing log here would be nice in that others could find it in the future through search engines.)

Also see https://github.com/gogits/gogs/issues/636

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/39626264-pushing-a-large-repo-with-http-leads-to-crash?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by typeless**
Not perfect but I think it's a reasonable solution.
For small request bodies, I suppose performance wouldn't be an issue.
For large ones, this seems to be a necessary evil.

Fixes #218

### CODE REVIEW

**Comment by typeless** [comment]
Any open issue for this?

File: routers/repo/http.go [hunk 1]
```diff
@@ -364,14 +364,23 @@ func serviceRPC(h serviceHandler, service string) {
 	}
 
 	if h.cfg.OnSucceed != nil {
-		input, err = ioutil.ReadAll(reqBody)
+		tmpfile, err := ioutil.TempFile("", "gogs")
```
Comment (strk) [inline routers/repo/http.go:367]: I hadn't found info about the permissions that TempFile would put on the file, I guess it should be checked

Comment (typeless) [reply routers/repo/http.go:367]: https://github.com/golang/go/blob/master/src/io/ioutil/tempfile.go#L55

Is this what you are looking for?

**Comment by strk** [comment]
No issue that I know of, please leave here a reference to the Gogs PR which in turn has more refs to Gogs issue and if you feel like it, also add a local issue

**Comment by lunny** [comment]
I don't think it's a good idea to use temporary file always since most time, the upload data is small data on git situation. For large files, lfs maybe a better choice and we have a merge-ready PR #122 .

**Comment by typeless** [comment]
@lunny I am genuinely open to a better solution but I am afraid that LFS cannot solve the problem if the repository in huge size consists of many small files rather than some big binary blobs.

Maybe someone else has experiences about how other similar projects deal with such issue?

**Comment by lunny** [comment]
I think this PR maybe resolve #218 . Could we chose memory or file according the content-size ?

**Comment by thibaultmeyer** [comment]
I think a max size must be defined (maybe on app.ini) to avoid app crash or other kind of strange issue if free space is not enough to store temp data.

**Comment by tboerger** [comment]
The current approach will slow down all writes, and that's sad

**Comment by lunny** [comment]
^
That's what I want to say. And most times, it will slow down.

**Comment by thibaultmeyer** [comment]
Id not possible to process incoming data as stream rather than just wait upload was completed ?

**Comment by lunny** [comment]
Maybe we can use this library to do something https://github.com/edsrzf/mmap-go

**Comment by strk** [comment]
A stream oriented approach would be best, but it isn't trivial to implement (me and @bkcsoft tried it with no much luck) - if I recall correctly it had to do with the way the code was written, that's to say a refactoring might make it easier to do stream-oriented. Want to give it another try ?

**Comment by lunny** [comment]
I think `mmap` is the correct method to resolve it. Keep the performance and resolve the large file bug.

**Comment by typeless** [comment]
Just took a glance and it seems a stream-oriented approach would be more probable if we can change the function signature of the `[OnSucceed](https://github.com/go-gitea/gitea/blob/master/routers/repo/http.go#L233)` from  `func(rpc string, input []byte)` to `func(rpc string, input io.Reader)`.

**Comment by typeless** [comment]
If Windows support is not an issue then `open(2)` with `O_TMPFILE` could be an option as well. ( I am not sure how well the status of cross-platform support of Gitea is)

**Comment by tboerger** [comment]
Gitea MUST work on Windows as well

**Comment by typeless** [comment]
@tboerger If that's the case, then `O_TMPFILE` and `mmap` would introduce some extra platform-dependent  complexities.

**Comment by strk** [comment]
I'm replying to this comment which I don't find on the web ui
anymore (deleted?):

On Tue, Nov 22, 2016 at 06:12:49PM -0800, typeless wrote:

> Just took a glance and it seems a stream-oriented approach would be more probable if we can change the function signature of the `[OnSucceed](https://github.com/go-gitea/gitea/blob/master/routers/repo/http.go#L233)` from  `func(rpc string, input []byte)` to `func(rpc string, input io.Reader)`.

This kind of refactoring did sound like a good idea, did you
hit a roadblock that made you remove the comment ?

**Comment by typeless** [comment]
@strk Yeah, I did delete the comment right after I realized that it's not just simply changing the function signature.  The problem is that we have to read though the stream *twice* and the reader is not necessarily seekable.

Need to take a look at the underlying mechanism of the origin of the request body (how the data is connected/copied to the io.Reader).

**Comment by lunny** [comment]
This is only because we maybe called `h.cfg.OnSucceed`. Or we can just let `cmd.Stdin = reqBody`.

**Comment by strk** [comment]
Maybe there's the possibility of using a Tee

**Comment by typeless** [comment]
> This is only because we maybe called h.cfg.OnSucceed. Or we can just let cmd.Stdin = reqBody.

@lunny Yeah, hopefully someone knows about why it has to be a two-pass operation in the first place.

**Comment by typeless** [comment]
> Maybe there's the possibility of using a Tee

@strk io.Pipe() + goroutines seems to be the straightforward way.  But I guess that would make the `Read` of Tee blocking if the other end of read in `OnSucceed` isn't running simultaneously.

**Comment by bkcsoft** [comment]
How about this then: (AFAIK) you can check the payload-size and for anything above `MAX_HTTP_PAYLOAD` would go into a TempFile ?

**Comment by bkcsoft** [comment]
About using TeeWriter, I've tried that without success... The good fix for this is to move `h.Cfg.OnSucced` to `.git/hooks/post-recieve` (where it belongs), but doing that would make said git-hook unavailable to the User/Admin (i.e. break backwards-compat).

**Comment by typeless** [comment]
@bkcsoft I have a PR which adds what you suggested but the value of `MAX_HTTP_PAYLOAD` is still a TBD.
Any suggestions on the value and the location of the constant definition?

I have lost the original repo during the process of transitioning the Github fork to Gitea from Gogs (Local repo is still there, though I am afraid it's impossible to restore the tracking for Github). I would have to either resend this PR or send another PR addressing the threshold after this getting merged.

Which means, I will probably leave the PR as-is and send follow-ups for `MAX_HTTP_PAYLOAD` if that's the agreed direction.

**Comment by strk** [comment]
if you're going to define a new constant use CamelCase please ..

**Comment by tboerger** [comment]
I have moved it to 1.1.0 for now.

**Comment by strk** [comment]
For the record, @Unkwon merged this in Gogs: https://github.com/gogits/gogs/issues/2960

**Comment by lunny** [comment]
A temporary method is to write to a template file in an another goroutine and write actions information to database. Every time write disk will slow the operations. So I don't think merge this directly is a good idea.

**Comment by typeless** [comment]
Is there any opportunity to handle this more gracefully (i.e. to avoid copying completely) if those operations are replaced with git library calls instead of fork/exec to `git`? Just saying.

**Comment by lunny** [comment]
I think @bkcsoft 's idea is the good direction.

> About using TeeWriter, I've tried that without success... The good fix for this is to move h.Cfg.OnSucced to .git/hooks/post-recieve (where it belongs), but doing that would make said git-hook unavailable to the User/Admin (i.e. break backwards-compat).

**Comment by bkcsoft** [comment]
@lunny using `post-recieve`-hook would be 2 PRs in this case

1. move the user-provided hook somewhere else (`hooks/gitea-post-recieve`?) and have `hooks/post-recieve` call that script
2. move this code into `hooks/post-recieve` after user-provided script is ran

**Comment by lunny** [comment]
I think this should be closed. And let's discuss in #218

**Pull request closed without merging**